### PR TITLE
[Tablet Orders] Update completion button based on flag, and navigation title based on size class

### LINKED ISSUE #12057: [Tablet Orders] Design tweaks
From review https://github.com/woocommerce/woocommerce-ios/pull/12033#pullrequestreview-1891221396
- [x] Remove `ProductSelectorView` divider between header and list:
⚠️ This seems to be added by default with SwiftUI navigation on iOS16+. Needs more research.
- [x] Add `ProductSelectorView` divider between header and list
- [x] Segmented control is truncated
We can use `.segmented` vs `.menu` built-in types based on size class & screen width
- [x] Adjust header in split view
- [x] `SearchBar` iPhone unnecessary padding and truncation
- [x] Scan button using wrong tint on dark mode

### PR DESCRIPTION
**Pull request by iamgabrielma**
Closes: #12057

## Description
This PR addresses 2 design tweaks:
* On iPhones, portrait: The product selector completion button currently displays the number of selected products when we select or deselect them. With the new designs this will only show either "Done" if no products have been selected, or "Add Products" text if any has been selected.
* On iPhones, portrait: The "selected products" text that was showing in the completion button, now has been moved to the navigation header.

Screenshot 2024-02-28 at 11 08 40

## Testing instructions
* With `.sideBySideViewForOrderForm` disabled, observe that the view behaves as before: "Add Product" shows in the navigation bar, and the button is updated with the amount of selected products.
* With `.sideBySideViewForOrderForm` enabled, observe that the view behaves as after: The navigation bar shows the number of selected products, and the button only switches between "Done" (if no products have been selected) or "Add Products" (if any has been selected).

| iPhone before | iPhone after |
|--------|--------|
| simulator_screenshot_1EBEB86D-5A87-4031-87A5-0EEEDD373097 | simulator_screenshot_26EEC54D-6CFA-436F-A28C-B78A2A3E3F09 | 

## Edge case in tablets:

There's one edge case that is not addressed in this PR: If the feature flag is enabled, and we run the app using split view where the product selector is narrow, but not narrow enough to use modal-on-modal, then we'll see only "Add Products" in the navigation title. This will not happen if we "move to that view", only when we initialize the product selector in that specific setup.

20240228 edge case tablet splitview narrow

This is because the `onChange` property that checks the current subview width and needs to notify the parent in order to update the navigation title is not called on initialization, but only when the property changes. Addressing this seems to be a better fit when we attempt to switch from the current usage of GeometryReader to specific adaptive views.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by joshheald** [comment]
> This is because the `onChange` property that checks the current subview width and needs to notify the parent in order to update the navigation title is not called on initialization, but only when the property changes. Addressing this seems to be a better fit when we attempt to switch from the current usage of GeometryReader to specific adaptive views.

FYI using `onAppear` to do the same as `onChange` is a useful and common approach – you need both, but it might fix this.

**Review by joshheald — APPROVED** [review]
No product selection count on iPad modal

I noticed that in a smaller iPad presentation, and on my iPhone however I get to it, both the title and the button say `Add Products`. This appears to be because the `isViewWidthNarrowerThanConstantRowWidth` property is never getting set.

No count on iPhone

Adding the `.onAppear` as I suggested in my last comment seems to fix this.

It'll need some more testing I think, when you've made that change... I just did a quick check.

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift [hunk 1]
```diff
@@ -79,9 +84,22 @@ struct ProductSelectorView: View {
         guard viewModel.totalSelectedItemsCount > 0 else {
             return Localization.doneButton
         }
-        return String.pluralize(viewModel.totalSelectedItemsCount,
-                                singular: configuration.doneButtonTitleSingularFormat,
-                                plural: configuration.doneButtonTitlePluralFormat)
+        guard ServiceLocator.featureFlagService.isFeatureFlagEnabled(.sideBySideViewForOrderForm) else {
+            return String.pluralize(viewModel.totalSelectedItemsCount,
+                                    singular: configuration.doneButtonTitleSingularFormat,
+                                    plural: configuration.doneButtonTitlePluralFormat)
```
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift]: can/should we use the `viewModel.selectProductsTitle` here too?

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/188">17.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift (hunk 1 above)
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift]: Yes, when I was extracting logic outside of the view I saw we have some duplication when populating the view, some of it comes from the configuration and then we modify it within the view, however, these same string operations can be found in the view model. Changed on https://github.com/woocommerce/woocommerce-ios/pull/12126/commits/0383efbc764fdcde6abd25f58d0aa8842c5cb517

**Comment by iamgabrielma** [comment]
Thanks for the review!

> Adding the .onAppear as I suggested in my last comment seems to fix this. It'll need some more testing I think, when you've made that change... I just did a quick check.

That seemed to fix it initially, but I found more edge cases mainly related to using a constant of 450 units as threshold, which works fine on iPhone is not enough for some bigger views, as well as what seems to be the SwiftUI view lifecycle, since checking the view's width `.onAppear` returns 0.0 in some instances unless we perform a DispatchQueue there, which adds more complexity.

Along with your `.onAppear` suggestion, I added an additional check for the size class that seems to fix all cases I tested: https://github.com/woocommerce/woocommerce-ios/pull/12126/commits/439c9a5c44cd1376a9e121eb43c74661b82b146d

**Pull request merged**